packages/dma-contracts/tasks/verify-deployment/index.ts

```diff
@@ -0,0 +1,279 @@
+import { ServiceRegistry } from '@typechain/index'
+import { color } from 'console-log-colors'
+import { task } from 'hardhat/config'
+
+import {
+  getServiceRegistry,
+  isInvalidAddress,
+  ServiceRegistryMaybe,
+  VerificationResult,
+} from '../common'
+const { red, yellow, green } = color
+
+function getValidationStatusString(verificationResult: VerificationResult): string {
+  const verificationPercentage: number = Math.round(
+    (verificationResult.totalValidated / verificationResult.totalEntries) * 100,
+  )
+
+  if (verificationPercentage < 30) {
+    return (
+      red(`${verificationPercentage.toFixed(2)}%`) +
+      ` (${verificationResult.totalValidated}/${verificationResult.totalEntries})`
+    )
+  } else if (verificationPercentage < 70) {
+    return (
+      yellow(`${verificationPercentage.toFixed(2)}%`) +
+      ` (${verificationResult.totalValidated}/${verificationResult.totalEntries})`
+    )
+  } else {
+    return (
+      green(`${verificationPercentage.toFixed(2)}%`) +
+      ` (${verificationResult.totalValidated}/${verificationResult.totalEntries})`
+    )
+  }
+}
+
+async function validateContracts(
+  config: { [key: string]: SystemConfigEntry },
+  serviceRegistry: ServiceRegistry,
+): Promise<VerificationResult> {
+  let totalEntries = 0
+  let totalValidated = 0
+
+  if (Object.keys(config).length === 0) {
+    return {
+      success: false,
+      totalEntries,
+      totalValidated,
+    }
+  }
+
+  for (const entryName in config) {
+    if (entryName === 'ServiceRegistry') {
+      continue
+    }
+
+    totalEntries++
+
+    const configEntry: SystemConfigEntry = config[entryName]
+    if (configEntry.serviceRegistryName) {
+      const registryAddress = await serviceRegistry.getRegisteredService(
+        configEntry.serviceRegistryName,
+      )
+
+      if (registryAddress === ADDRESS_ZERO) {
+        console.log(`${entryName}: ❌ (not deployed)`)
+        continue
+      }
+
+      if (registryAddress.toLowerCase() !== configEntry.address.toLowerCase()) {
+        console.log(
+          `${entryName}: ❌ (address mismatch ${configEntry.address} <--> ${registryAddress})`,
+        )
+        continue
+      }
+    }
+
+    totalValidated++
+
+    console.log(`${entryName}: ✅ (${configEntry.address})`)
+  }
+
+  return {
+    success: totalEntries === totalValidated,
+    totalEntries,
+    totalValidated,
+  }
+}
+
+async function validateDependencies(config: {
+  [key: string]: ConfigEntry
+}): Promise<VerificationResult> {
+  let totalEntries = 0
+  let totalValidated = 0
+
+  if (Object.keys(config).length === 0) {
+    return {
+      success: false,
+      totalEntries,
+      totalValidated,
+    }
+  }
+
+  for (const entryName in config) {
+    totalEntries++
+
+    const configEntry: ConfigEntry = config[entryName]
+
+    if (isInvalidAddress(configEntry.address)) {
+      console.log(`${entryName}: ❌ (not configured)`)
+      continue
+    }
+
+    totalValidated++
```

Comment:
or even we could do:
```
const validated = Object.entries(config).filter(([key, value] ) => { 
if (isInvalidAddress(configEntry.address)) {
      console.log(`${entryName}: ❌ (not configured)`)
      return false
    }
    return true
})
```

Then we can remove mutating states of the variables.

Reply:
If I understand correctly, then your idea is to count the number of validated agains the total number of entries in config, right? If so, I like it better, thanks!